src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/Deleter.cs

```diff
@@ -0,0 +1,113 @@
+            _conformanceProvider = conformanceProvider;
+            _fhirDataStore = fhirDataStore;
+            _searchService = searchService;
+            _resourceIdProvider = resourceIdProvider;
+        }
+
+        public async Task<ResourceKey> DeleteAsync(DeleteResourceRequest request, CancellationToken cancellationToken)
+        {
+            var key = request.ResourceKey;
+
+            if (!string.IsNullOrEmpty(key.VersionId))
+            {
+                throw new MethodNotAllowedException(Core.Resources.DeleteVersionNotAllowed);
+            }
+
+            string version = null;
+
+            switch (request.DeleteOperation)
+            {
+                case DeleteOperation.SoftDelete:
+                    var emptyInstance = (Resource)Activator.CreateInstance(ModelInfo.GetTypeForFhirType(request.ResourceKey.ResourceType));
+                    emptyInstance.Id = request.ResourceKey.Id;
+
+                    ResourceWrapper deletedWrapper = _resourceWrapperFactory.CreateResourceWrapper(emptyInstance, _resourceIdProvider, deleted: true, keepMeta: false);
+
+                    bool keepHistory = await _conformanceProvider.Value.CanKeepHistory(key.ResourceType, cancellationToken);
+
+                    UpsertOutcome result = await _fhirDataStore.UpsertAsync(new ResourceWrapperOperation(deletedWrapper, true, keepHistory, null, false), cancellationToken);
+
+                    version = result?.Wrapper.Version;
+                    break;
+                case DeleteOperation.HardDelete:
+                case DeleteOperation.PurgeHistory:
+                    await _fhirDataStore.HardDeleteAsync(key, request.DeleteOperation == DeleteOperation.PurgeHistory, cancellationToken);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(request));
+            }
+
+            return new ResourceKey(key.ResourceType, key.Id, version);
+        }
+
+        public async Task<int> DeleteMultipleAsync(ConditionalDeleteResourceRequest request, CancellationToken cancellationToken)
+        {
+            (IReadOnlyCollection<SearchResultEntry> matchedResults, string ct) = await _searchService.ConditionalSearchAsync(request.ResourceType, request.ConditionalParameters, request.MaxDeleteCount, cancellationToken);
+
+            int itemsDeleted = 0;
+            bool deleteAll = request.MaxDeleteCount < 0;
+
+            // Delete the matched results...
+            while (matchedResults.Any() || !string.IsNullOrEmpty(ct))
+            {
+                var resultsToDelete = deleteAll ? matchedResults : matchedResults.Take(request.MaxDeleteCount - itemsDeleted);
+                foreach (SearchResultEntry item in resultsToDelete)
+                {
+                    var result = await DeleteAsync(new DeleteResourceRequest(request.ResourceType, item.Resource.ResourceId, request.DeleteOperation), cancellationToken);
```

Comment:
I believe that in this case, if these are all soft deletes, we can use MergeAsync method and delete all resources in a single database operation. That would be great in terms of performance.